### DISCUSSION TITLE
Synching multiple RedPitaya boards

### DISCUSSION THREAD

**Discussion by michaelcroquette** [lneuhaus/pyrpl#519] [Show and tell]
To overcome the limitations of having a finite number of inputs, outputs and IQ module and one card, we came up with a solution to synchronize two Red Pitaya boards on one external clock signal and then used a share trigger signal so the IQ modules are synced in phase over the two boards. This solution can be used to synchronize up to 4 boards in theory but we only tested it for two. 

### Clock signal generation 

We need an external clock source, we use a VersaClock 6E Programmable Clock Generator (5P49V60) mounted on a SparkFun board. The board is programmable via I2C code using an Arduino library (Arduino code can be found [here](https://github.com/michaelcroquette/SparkFun_Clock_5P49V60_Arduino_Library.git)). It can generate four 125 MHz clock signals with LVDS level (the standard needed for the Red Pitayas clock signal). You then have to connect the Red Pitaya clock inputs to the board and you can synchronize up to 4 Red Pitayas (In our case we only needed to synchronize two). 

![RP_synchro (1)](https://github.com/user-attachments/assets/5ef72373-c301-457c-bc54-14ee442efc10)
Clock generator connections.

### Red Pitaya Board modification

In order for the Red Pitaya board to use the external clock signal and not its own crystal clock signal, some soldering is needed. Unfortunately the official doc page is not responding anymore but a description of the procedure can be found here : https://forum.redpitaya.com/viewtopic.php?t=24704

### IQ module synchronization

**Warning : For now this feature is only implemented in the "modif_fpga_pfd" branch of pyrpl. On this branch, the pfd output option of the IQ module uses the CORDIC algorithm to give a precise estimation of the unwrap phase (more details [here](https://arxiv.org/abs/2310.00086)) .  For FPGA resource reason, there is no IIR module but 4 PID modules on this branch.** 

The last step is to make sure that the demodulation modules all start with the same phase. Even if the clocks are synchronized, the relative phase between the IQ module depends on the timing at which they were started. So we add a connector between two digital Input/Output of the RedPitayas, and when one of the RedPitaya starts its IQ module it will send a signal to the other so that their internal reference always have the same phase relationship. This was implemented in the FPGA code in a simple way : the IQ will start when the DIO0_P (ext. trig) pin of the RP board is up. So by connecting the two pins of the two boards together and setting one as output and the other as input using the housekeeping module (hks). 

output setup
![image](https://github.com/user-attachments/assets/2dc205d9-b32b-40d6-ad75-ea919527a70c)

input setup 
![image](https://github.com/user-attachments/assets/cdd306c3-59e4-469a-a443-5dc562d54466)

Once the pins are connected, enabling expansion_P0 will start the IQs simultaneously over the two boards. 
![image](https://github.com/user-attachments/assets/ab1fc0b2-aff4-4ff9-9158-2c6751f2b4e7)

Alternatively you can use any external trigger signal to synchronize more than two RPs. 
I hope this was clear and helpfuf, don't hesitate to ask more questions and give us feedback if you try to replicate this method !

**Comment by michaelcroquette** [comment]
I'm just adding a link to the original discussion #350 for those who would want to explore other way of synchronising RPs.

**Comment by JanWuest** [comment]
Hi,
I am currently trying to synchronize two IQ modules on two seperate Red Pitayas. I have read some of the Issues here but I can't find a solution.

## Hardware:
- Two Red Pitayas with ext. Clock modification (swapped Resistors R25 and R26 to R23 and R24 on the board) 
- Two Click Shield Boards, setup like [shown in the documentation](https://redpitaya.readthedocs.io/en/latest/appsFeatures/examples/multiboard_sync/mutliboard-3-click_shield_acq.html) 

This leads to both Red Pitayas using the same clock and the trigger signal is forwarded through the click shield aswell.

## Software:
- I have tried it first with the 0.9.5.0 and with the 0.9.7 release of pyrpl. Have to check 0.9.8 aswell. But here it was stated that I need to use the [modif-fpga-branch](https://github.com/pyrpl-fpga/pyrpl/tree/modif_fpga_pfd) which I did aswell
- I have checked and my pyrpl version has 4 PID modules and no IIR like you said.
- I use python 3.13 with spyder to call pyrpl and then use a combination of GUI and code via python. 


## What works so far:
1. I can set the same Frequency for both iq modules on the two red pitayas. And when I output the signal on one and connect it to the input of the other I can compare them in the scope and they are phase stable. But when I click "Synchronize IQ" it synchs the Frequency Generators of the IQ Modules of the same board but not on the other one. The start phase (or the pahse difference) between both boards is therefore random. 
2. I can use the hk module to set the expansion P0 pin on both Red Pitayas as an input and the expansion N0 pin on the first Red Pitaya as an output. When I use the asg modules and set the trigger source to "ext_positive_edge" the asg modules start wih the given start phases, as soon as I set N0 to "true". The Trigger Signal seems to work atleast for the asg modules. 

## What doesn't work

> So we add a connector between two digital Input/Output of the RedPitayas, and when one of the RedPitaya starts its IQ module it will send a signal to the other so that their internal reference always have the same phase relationship. This was implemented in the FPGA code in a simple way : the IQ will start when the DIO0_P (ext. trig) pin of the RP board is up. 

You described that the IQ module will start as soon as P0 is up. When I use the refresh button in the hk module I can see that P0 goes "up" as soon as I set N0 to "true". Also the Trig_out of the Second Clickshield is following N0 of the first board (I checked with an oscilloscope) so I am pretty sure that the P0 port of the red Pitaya is set correctly. 

The IQ modules start running as soon as I enter a frequency and amplitude value. And don't wait for the trigger pin to go up. Maybe my understanding of "starting the IQ module" is wrong. But I have not found another option to "start" the IQ module.

## What I wish it would do:
Some form of "starting" or "synching" the frequency generator in the IQ modules of two different Red Pitayas board using the trigger input pin P0. Either by clicking "synch IQs" on the first Pyrpl instance or manually setting a trigger pin. 

I would be happy if you could help me :) 

Thank you for your work, pyrpl is a great tool for our lab. 

Best Jan

**Reply by michaelcroquette** [reply]
Hi,
Cool that you are trying Click Shield ! I haven't tried them myself. I think the issue was coming from me. You absolutely need to use the modif_fpga_pfd branch, for now it's the only one where the change is implemented. But I just saw that I had not pulled the most recent commits from main so the loading of the fpga image was not working properly. I pulled them and did a quick test and it's working for me now. 

Normally if you have the correct fpga loaded, the iq module should not generate any output until expansion_P0 is high. 

To make sure you are loading the fpga properly, run `Pyrpl(..., reloadfpga=True, loglevel='debug')`.

You should see something like this in the output : 

`DEBUG:pyrpl.redpitaya:cat /root/.version: cat /root/.version
2.07
root@rp-f0cc02:~#
2.07G:pyrpl.redpitaya:OS version:
DEBUG:pyrpl.redpitaya:reloadfpga=True: FPGA will be reloaded
DEBUG:pyrpl.redpitaya:Base directory: C:\Users\croquette\GitHub\pyrpl\pyrpl
DEBUG:pyrpl.redpitaya:Checking for FPGA binfile at: C:\Users\croquette\GitHub\pyrpl\pyrpl\fpga/red_pitaya.bin
INFO:pyrpl.redpitaya:Found FPGA binfile at: C:\Users\croquette\GitHub\pyrpl\pyrpl\fpga/red_pitaya.bin
INFO:pyrpl.redpitaya:Found DTBO file at: C:\Users\croquette\GitHub\pyrpl\pyrpl\fpga/red_pitaya.dtbo
DEBUG:pyrpl.sshshell:
DEBUG:pyrpl.sshshell:^C

root@rp-f0cc02:~#
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:killall monitor_server
monitor_server: no process found
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:rw
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:mkdir -p //opt//pyrpl//
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:^C

root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:killall monitor_server
monitor_server: no process found
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:killall nginx
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:systemctl stop redpitaya_nginx
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:/opt/redpitaya/sbin/overlay.sh pyrpl //opt//pyrpl//fpga.bit.bin //opt//pyrpl//fpga.dtbo

DEBUG:pyrpl.sshshell:
root@rp-f0cc02:~#
DEBUG:pyrpl.sshshell:cat /tmp/update_fpga.txt
Time taken to load BIN is 165.000000 Milli Seconds
BIN FILE loaded through FPGA manager successfully
FPGA md5sum: 2a695e1314f24e8e94b06c2868ca94f2  /opt/pyrpl/fpga.bit.bin
Thu Mar 19 07:16:12 PM UTC 2026`

**Reply by michaelcroquette** [reply]
Let me know how it goes !

**Reply by michaelcroquette** [reply]
Out of curiosity, from which lab are you ? Please cite our article if you are using pyrpl for your research  https://pubs.aip.org/aip/rsi/article-abstract/95/3/033003/3269904/Python-Red-Pitaya-Lockbox-PyRPL-An-open-source?redirectedFrom=fulltext:
